### PR TITLE
Mt7921e devel - MTK internal SQC fixes - v2

### PR DESCRIPTION
**Pull request by objelf**
Hi Lore, 

Please help review if there is any problem.

If the patch is fine to you, then I would add your Reviewed-by: tag and send them out to the upstream.

thanks,
Sean

### CODE REVIEW

File: drivers/net/wireless/mediatek/mt76/mt76_connac.h [hunk 1]
```diff
@@ -134,8 +134,12 @@ mt76_connac_pm_unref(struct mt76_connac_pm *pm)
 static inline bool
 mt76_connac_skip_fw_pmctrl(struct mt76_phy *phy, struct mt76_connac_pm *pm)
 {
+	struct mt76_dev *dev = phy->dev;
 	bool ret;
 
+	if (mt76_is_mmio(dev) && dev->token_count)
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac.h]: @objelf I do not think we need the mt76_is_mmio() check is token_count is increased/decreased by mmio device only. I will remove the check merging the patch

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 2]
```diff
@@ -875,10 +875,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
-	mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta, info->enable);
-	if (info->enable && info->sta)
-		mt76_connac_mcu_sta_tlv(phy, skb, info->sta, info->vif,
-					info->rcpi);
+	if (info->sta) {
+		mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta,
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: @objelf mt76_connac_mcu_add_sta_cmd() is run by mt7663 driver with info->sta = NULL and info->enable = NULL in mt7615_remove_interface(). Can you please double check the mt76_connac_mcu_sta_basic_tlv() is not necessary in this case (mt7663 device)?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 3]
```diff
@@ -1318,8 +1319,21 @@ int mt76_connac_mcu_uni_add_bss(struct mt76_phy *phy,
 				.tag = cpu_to_le16(UNI_BSS_INFO_HE_BASIC),
 				.len = cpu_to_le16(sizeof(struct bss_info_uni_he)),
 			},
+			.bss_color = {
+				.tag = cpu_to_le16(UNI_BSS_INFO_BSS_COLOR),
+				.len = cpu_to_le16(sizeof(struct bss_info_uni_bss_color)),
+				.enable = 0,
+				.bss_color = 0,
+			},
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: @objelf can you please confirm mt7663 just ignore this field?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c [hunk 4]
```diff
@@ -218,7 +218,7 @@ void mt76_connac_mcu_beacon_loss_iter(void *priv, u8 *mac,
 	if (!(vif->driver_flags & IEEE80211_VIF_BEACON_FILTER))
 		return;
 
-	ieee80211_beacon_loss(vif);
+	ieee80211_connection_loss(vif);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: @objelf is it true for mt7663 fw as well?

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c [hunk 5]
```diff
@@ -13,6 +13,9 @@ int mt76_connac_pm_wake(struct mt76_phy *phy, struct mt76_connac_pm *pm)
 	if (mt76_is_usb(dev))
 		return 0;
 
+	if (test_bit(MT76_STATE_SUSPEND, &phy->state))
+		return 0;
+
 	cancel_delayed_work_sync(&pm->ps_work);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c]: @objelf is it enough to just return after cancel_delayed_work_sync() if the device is not running (MT76_STATE_RUNNING not set)? Doing so we can drop the MT76_STATE_WOW patch, what do you think?

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 6]
```diff
@@ -1301,24 +1302,30 @@ mt7921_mac_reset(struct mt7921_dev *dev)
...
+	cancel_delayed_work_sync(&pm->ps_work);
+	cancel_work_sync(&pm->wake_work);
+
 	mutex_lock(&dev->mt76.mutex);
 	for (i = 0; i < 10; i++) {
+		__mt7921_mcu_drv_pmctrl(dev);
+		clear_bit(MT76_STATE_PM, &dev->mphy.state);
+		dev->pm.stats.last_wake_event = jiffies;
+		dev->pm.stats.last_doze_event = jiffies;
+
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: @objelf why not moving dev->pm.stats.last_wake_event/dev->pm.stats.last_doze_event initialization and clear_bit() inside __mt7921_mcu_drv_pmctrl()?

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c [hunk 7]
```diff
@@ -1339,13 +1346,16 @@ void mt7921_mac_reset_work(struct work_struct *work)
...
+	mt76_connac_power_save_sched(&dev->mt76.phy, &dev->pm);
 }
 
 void mt7921_reset(struct mt76_dev *mdev)
 {
 	struct mt7921_dev *dev = container_of(mdev, struct mt7921_dev, mt76);
 
-	queue_work(dev->mt76.wq, &dev->reset_work);
+	if (!test_bit(MT76_HW_RESET, &dev->mphy.state))
+		queue_work(dev->mt76.wq, &dev->reset_work);
```
Comment (LorenzoBianconi) [inline drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: @objelf do we really need this check? A similar order is already performed by queue_work() and mt76 is an ordered workqueue

**Comment by LorenzoBianconi** [comment]
> Hi Lore,
> 
> Please help review if there is any problem.
> 
> If the patch is fine to you, then I would add your Reviewed-by: tag and send them out to the upstream.
> 
> thanks,
> Sean

Hi Sean, 

I merged most the patches and I added some comments in the other ones (most of them are trivial to fix)

Regards,
Lorenzo

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 3 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: the UNI_BSS_INFO_BSS_COLOR only supports 80211ax chipset

Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: the UNI_BSS_INFO_BSS_COLOR only supports 80211ax chipset

File: drivers/net/wireless/mediatek/mt76/mt76_connac.h (hunk 1 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac.h]: ok, thanks.

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 4 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: The change should be dropped. let's focus on mt7921 first.

File: drivers/net/wireless/mediatek/mt76/mt7921/mac.c (hunk 6 above)
Comment (objelf) [reply drivers/net/wireless/mediatek/mt76/mt7921/mac.c]: it seems doable. let's do it in the next version.

File: drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c (hunk 3 above)
Comment (LorenzoBianconi) [reply drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c]: what I mean is, is bss_info_uni_bss_color element just skipped by 7663 fw or does it report an error? in the first case the patch is fine

**Comment by objelf** [comment]
close in order to send  the connection monitor patch

**Pull request closed without merging**